### ISSUE TITLE
Mapping failed when PropertyMap source type is an interface of source

### ISSUE THREAD

**Issue by chhsiao90** [modelmapper/modelmapper#202]
Original issue is at #144 comment : [144#issuecomment-279717769](https://github.com/jhalterman/modelmapper/issues/144#issuecomment-279717769)
And problem I had addressed at [144#issuecomment-279963373](https://github.com/jhalterman/modelmapper/issues/144#issuecomment-279963373)

Minimum reproducer as following code snippet
the testMapping2 will failed

```java
package org.modelmapper.bugs;

import static org.testng.Assert.assertEquals;

import org.modelmapper.AbstractTest;
import org.modelmapper.PropertyMap;
import org.testng.annotations.Test;

@Test
public class GH202 extends AbstractTest {

    public void testMapping1() {
        modelMapper.addMappings(new Mappings1());

        DestinationModel dest = modelMapper.map(new SourceModel(), DestinationModel.class);

        assertEquals(dest.getName(), "Bob Alice");
        assertEquals(dest.getNestedModel().getAge(), "99");
    }

    public void testMapping2() {
        modelMapper.addMappings(new Mappings2());

        DestinationModel dest = modelMapper.map(new SourceModel(), DestinationModel.class);

        assertEquals(dest.getName(), "Bob Alice");
        assertEquals(dest.getNestedModel().getAge(), "99");
    }

    static class DestinationModel {
        private String name;
        private NestedDestinationModel nestedModel;

        public String getName() {
            return name;
        }

        public void setName(final String name) {
            this.name = name;
        }

        public NestedDestinationModel getNestedModel() {
            return nestedModel;
        }

        public void setNestedModel(final NestedDestinationModel nestedModel) {
            this.nestedModel = nestedModel;
        }
    }

    static class NestedDestinationModel {
        private String age;

        public String getAge() {
            return age;
        }

        public void setAge(final String age) {
            this.age = age;
        }
    }

    static class SourceModel implements MappingInterface {
        public String getName() {
            return "Bob Alice";
        }

        public String getAge() {
            return "99";
        }
    }

    interface MappingInterface {
        String getName();
        String getAge();
    }

    class Mappings1 extends PropertyMap<SourceModel, DestinationModel> {
        protected void configure() {
            map().setName(source.getName());
            map().getNestedModel().setAge(source.getAge());
        }
    }

    class Mappings2 extends PropertyMap<MappingInterface, DestinationModel> {
        protected void configure() {
            map().setName(source.getName());
            map().getNestedModel().setAge(source.getAge());
        }
    }
}
```

**Comment by chhsiao90** [comment]
@DouglasBrunner - I think mapping depends on the interface of source is dangerous.
Because source can implement multiple interfaces, and ModelMapper may found multiple PropertyMaps that match the source's interface.

For example,
we had source with type
```
class SourceType implement InterfaceA, InterfaceB {
 ...
}
```
And defined PropertyMap for ModelMapper as following
```
modelMapper.addMappings(new PropertyMap<InterfaceA, DestType> {
  ...
});
modelMapper.addMappings(new PropertyMap<InterfaceB, DestType> {
  ...
});
```

What should ModelMapper do when it tried map SourceType to DestType?
Use PropertyMap<InterfaceA, DestType> or PropertyMap<InterfaceB, DestType>

For similar situation,
I think maybe provide a new function on ```ModelMapper``` to let user can pass sourceType explicitly.
like ```public <D> D map(Object source, Type destinationType, Class<?> sourceType)``` may solve the problem.

@jhalterman - What do you think!

**Comment by DouglasBrunner** [comment]
I see what you're saying, but isn't it the case that Java won't allow any ambiguous method signatures when implementing multiple interfaces?

I'll try to create an example this weekend when I have time to better explain what I mean.

**Comment by chhsiao90** [comment]
Just found it related to #13 and had a PR #160 for it

**Comment by chhsiao90** [comment]
Closed by #209

**Comment by chhsiao90** [comment]
The TypeMap inheritance had been release on ModelMapper v1.0.0, please give it a try!
